src/v/cloud_storage/topic_manifest.cc

```diff
@@ -12,21 +12,193 @@
+    using key_string = ss::basic_sstring<char, uint32_t, 31>;
+    bool StartObject() {
+        switch (_state) {
+        case expectManifestStart:
+            _state = expectKey;
+            return true;
+        case expectKey:
+        case expectValue:
+            return false;
+        }
+    }
+
+    bool Key(const char* str, rapidjson::SizeType length, bool /*copy*/) {
+        switch (_state) {
+        case expectKey:
+            _key = key_string(str, length);
+            _state = expectValue;
+            return true;
+        case expectManifestStart:
+        case expectValue:
+            return false;
+        }
+    }
+
+    bool String(const char* str, rapidjson::SizeType length, bool /*copy*/) {
+        std::string_view sv(str, length);
+        switch (_state) {
+        case expectValue:
+            if ("namespace" == _key) {
+                _namespace = model::ns(ss::sstring(sv));
+            } else if ("topic" == _key) {
+                _topic = model::topic(ss::sstring(sv));
+            } else if ("compression" == _key) {
+                model::compression compression_res;
+                if (boost::conversion::try_lexical_convert(
+                      sv, compression_res)) {
+                    _properties.compression = compression_res;
+                } else {
+                    return false;
+                }
+            } else if ("cleanup_policy_bitflags" == _key) {
+                model::cleanup_policy_bitflags cleanup_policy_bitflags_res;
+                if (boost::conversion::try_lexical_convert(
+                      sv, cleanup_policy_bitflags_res)) {
+                    _properties.cleanup_policy_bitflags
+                      = cleanup_policy_bitflags_res;
+                } else {
+                    return false;
+                }
+            } else if ("compaction_strategy" == _key) {
+                model::compaction_strategy compaction_strategy_res;
+                if (boost::conversion::try_lexical_convert(
+                      sv, compaction_strategy_res)) {
+                    _properties.compaction_strategy = compaction_strategy_res;
+                }
+            } else if ("timestamp_type" == _key) {
+                model::timestamp_type timestamp_type_res;
+                if (boost::conversion::try_lexical_convert(
+                      sv, timestamp_type_res)) {
+                    _properties.timestamp_type = timestamp_type_res;
+                }
+            } else {
+                return false;
+            }
+            _state = expectKey;
+            return true;
+        case expectManifestStart:
+        case expectKey:
+            return false;
+        }
+    }
+
+    bool Uint(unsigned u) { return Uint64(u); }
+
+    bool Uint64(uint64_t u) {
+        switch (_state) {
+        case expectValue:
+            if ("version" == _key) {
+                _version = static_cast<topic_manifest_version>(u);
```

Comment:
I think that `u` needs to be validated before casting in order to avoid undefined behavior if `u` is out-of-bounds / doesn't correspond to a enum value.

Reply:
IIUC `topic_manifest_version` has a scoped enumeration with a fixed underlying type `int32_t`
> `enum class topic_manifest_version : int32_t`

IIUC from https://wiki.sei.cmu.edu/confluence/display/cplusplus/INT50-CPP.+Do+not+cast+to+an+out-of-range+enumeration+value, it's ok not to validate it before casting and we should expected `int32_t` value

Reply:
I think that if `u` does not match an enumeration value in the scoped enumeration then the static cast is undefined behavior.